### PR TITLE
Add an alternative block appender when the container doesn't support the default block

### LINKED ISSUE #8589: Default Appender that is not Paragraph Based
In some cases with nested blocks, the default block appender with the `+` and recent blocks is not the right solution. We should offer an alternative appender that block authors can use, which would look just like a `+` button in the nested group.

Related: #6569 #7573.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #8589

This PR adds a new block appender that trigger only when the parent block doesn't support the default block. (paragraph).

To test it, try registering this block (copy/paste to the console)

```js
wp.blocks.registerBlockType('test/container', {
	title: 'container',
	category: 'common',
	description: 'container',
	icon: 'yes',

	edit() {
		return wp.element.createElement(wp.editor.InnerBlocks, {
			allowedBlocks: ['core/image', 'core/gallery']
		});
	},

	save() {
		return wp.element.createElement(wp.editor.InnerBlocks.Content);
	},
})
```

### CODE REVIEW

**Comment by jasmussen** [comment]
I love it. 

GIF:

new appender

Can you make it so it has a dashed outline like https://user-images.githubusercontent.com/1204802/44717667-f4f46b80-aabd-11e8-890d-d86f19cf3cea.png? That sort of implies where the boundaries of the inserter block will be.

Also, can you make it so the entire block is clickable? Sort of like how it is in the gallery?

gallery

Nice work! 👍 👍

**Comment by chrisvanpatten** [comment]
I love this!

I'd also kill for filters…

1. a filter to say "use the alternate block appender all the time" — useful if you're dealing with a post type that is primarily about layout, not textual content (I can think of a few cases where I'd use this)
2. a filter to replace the appender with something custom. You could use it to build some super cool contextual inserter experiences specific to a particular post type.

Overall, awesome and super excited for this.

**Review by tofumatt — CHANGES_REQUESTED** [review]
I noticed that I can still insert the default block even though it's not listed as an allowed type:



I guess that's true of the existing implementation, right? Do we consider that a bug?

Anyway, code looks mostly good, just some minor tweaks. Tested in IE 11 and seems good there too; I wanted to test there because some flex stuff can be weird there, but it all seemed good.

Might be a good case for an E2E test as this is might not be something routinely tested by us in development.

-----

My only issue is the block appender seems like it should have a larger click target:



It seems annoying to have to move my mouse to the centre; that's a tiny click target and we could expand it to the whole area.

File: packages/editor/src/components/block-list-appender/index.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	if ( isGroupedByLayout ) {
+		defaultLayout = layout;
+	}
+
+	if ( canInsertDefaultBlock ) {
+		return (
+			<IgnoreNestedEvents childHandledEvents={ [ 'onFocus', 'onClick', 'onKeyDown' ] }>
+				<DefaultBlockAppender
+					rootClientId={ rootClientId }
+					lastBlockClientId={ last( blockClientIds ) }
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply packages/editor/src/components/block-list-appender/index.js]: FWIW, last also avoids dealing with specific cases (an empty array). I mean `([])[-1]` does return undefined but I don't know if I trust it to not raise errors in all JavaScript contexts

**Comment by youknowriad** [comment]
@jasmussen I tried tweaking the design as suggested, feel free to polish further.

**Comment by chrisvanpatten** [comment]
@youknowriad I think a post-type level setting would be a nice alternative over a filter too. The main thing is that the default paragraph appender isn't always the best user interface when dealing with certain types of content, even in cases where the paragraph is an allowed block, and it would help to allow developers to be explicit about that rather than depending on Gutenberg to infer it.

**Comment by youknowriad** [comment]
> I noticed that I can still insert the default block even though it's not listed as an allowed type:

This is definitely a bug, good catch, I created a specific issue for it https://github.com/WordPress/gutenberg/issues/10186

File: packages/editor/src/components/block-list-appender/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (youknowriad) [reply packages/editor/src/components/block-list-appender/index.js]: I'm keeping it as a personal preference for me. But let me know if you feel strongly about it.

**Comment by jasmussen** [comment]
I pushed a small tweak to the visuals:

tweaks

File: test/e2e/specs/container-blocks.test.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		const paragraphToAdd = `
+			<!-- wp:paragraph -->
+			<p>added paragraph</p>
+			<!-- /wp:paragraph -->
+			`;
+		await insertBlock( blockName );
+		await switchToEditor( 'Code' );
+		await page.$eval( '.editor-post-text-editor', ( element, _paragraph, _blockSlug ) => {
+			const blockDelimiter = `<!-- /wp:${ _blockSlug } -->`;
+			element.value = element.value.replace( blockDelimiter, _paragraph + blockDelimiter );
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply test/e2e/specs/container-blocks.test.js]: (I know this is unrelated to your patch, but I'ma make the change because it's nicer.)

**Review by tofumatt — APPROVED** [review]
Tested locally and the bigger click target is *much* nicer, thanks!

Looks good to me! 👍

**Comment by youknowriad** [comment]
Thanks for the tweaks :)

**Comment by chrisvanpatten** [comment]
@youknowriad @jasmussen Would it make sense for this new appender design to apply to the Gallery image appender as well? I could see it being nice to have the same design since they are similar UIs, same icons, etc. Happy to open a PR if so :)

**Comment by jasmussen** [comment]
> @chrisvanpatten Yes, it does. I think the component is different but the design should definitely match. Some explorations are happening here #9682

> @youknowriad Oh cool, I'll let those explorations proceed then! 👍

I actually think those explorations, although on a PR, have somewhat stalled. So Chris if you have an ocean of bandwidth and desire to contribute, I can offer you an emoji medal for it, and I would suggest you work off of the mockups present here: https://github.com/WordPress/gutenberg/pull/9682#issuecomment-425435709

An interim step can be as little as the visual styles, and a bigger step could be adding the additional button. But just to be clear, don't let that pull request block you from making one yourself!

File: packages/editor/src/components/inserter/index.js [hunk 3]
```diff
@@ -88,26 +88,31 @@ class Inserter extends Component {
...
 			getInserterItems,
+			getBlockOrder,
 		} = select( 'core/editor' );
 		const insertionPoint = getBlockInsertionPoint();
-		const { rootClientId } = insertionPoint;
+		const parentId = rootClientId || insertionPoint.rootClientId;
 		return {
 			title: getEditedPostAttribute( 'title' ),
-			insertionPoint,
+			insertionPoint: {
```
Comment (aduth) [inline packages/editor/src/components/inserter/index.js]: We should almost never want to create a new object to be returned as a value of `mapSelectToProps` return value object.

See: #11028

**Pull request merged**